akka-persistence-typed/src/main/scala/akka/persistence/typed/javadsl/CommandHandler.scala

```diff
@@ -0,0 +1,91 @@
+  def apply(ctx: ActorContext[Command], state: State, command: Command): Effect[Event, State]
+}
+/**
+ * SAM for reacting on signals
+ *
+ * Used with [[CommandHandlerBuilder]] to setup the behavior of a [[PersistentBehavior]]
+ */
+@FunctionalInterface
+trait CommandToEffect[Command, MsgCommand <: Command, Event, State] {
+  /**
+   * @return An effect for the given command
+   */
+  def apply(ctx: ActorContext[Command], state: State, command: MsgCommand): Effect[Event, State]
+}
+
+object CommandHandlerBuilder {
+  def builder[Command, Event, State](rootCommandClass: Class[Command]): CommandHandlerBuilder[Command, Event, State] =
+    new CommandHandlerBuilder[Command, Event, State](rootCommandClass)
+}
+
+final class CommandHandlerBuilder[Command, Event, State] @InternalApi private[persistence] (rootCommandClass: Class[Command]) {
+
+  private final case class CommandHandlerCase(predicate: Command ⇒ Boolean, handler: CommandToEffect[Command, Command, Event, State])
+
+  private var cases: List[CommandHandlerCase] = Nil
+
+  private def addCase(predicate: Command ⇒ Boolean, handler: CommandToEffect[Command, Command, Event, State]): Unit = {
+    cases = CommandHandlerCase(predicate, handler) :: cases
+  }
+  /**
+   * Match any command which the given `predicate` returns true for
+   */
+  def matchCommand(predicate: FI.TypedPredicate[Command], commandToEffect: CommandToEffect[Command, Command, Event, State]): CommandHandlerBuilder[Command, Event, State] = {
+    addCase(cmd ⇒ predicate.defined(cmd), commandToEffect)
+    this
+  }
+
+  def matchCommand[C <: Command](commandClass: Class[C], commandToEffect: CommandToEffect[Command, C, Event, State]): CommandHandlerBuilder[Command, Event, State] = {
+    if (!rootCommandClass.isAssignableFrom(commandClass))
+      throw new IllegalArgumentException(s"Class ${commandClass.getName} is not a subtype of ${rootCommandClass.getName}")
+    addCase(cmd ⇒ commandClass.isAssignableFrom(cmd.getClass), commandToEffect.asInstanceOf[CommandToEffect[Command, Command, Event, State]])
+    this
+  }
+
+  /**
+   * Builds a Command Handler and resets this builder
+   */
+  def build(): CommandHandler[Command, Event, State] = {
+    val builtCases = cases.reverse.toArray
+    cases = Nil
+    new CommandHandler[Command, Event, State] {
+      override def apply(ctx: ActorContext[Command], state: State, command: Command) = {
+        var idx = 0
+        var effect: OptionVal[Effect[Event, State]] = OptionVal.None
+
+        while (idx < builtCases.length && effect.isEmpty) {
+          val curr = builtCases(idx)
+          if (curr.predicate(command)) {
+            effect = OptionVal.Some(curr.handler.apply(ctx, state, command))
+          }
+          idx += 1
+        }
+
+        effect match {
+          case OptionVal.None    ⇒ throw new MatchError(s"No match found for command of type [${command.getClass.getName}]")
+          case OptionVal.Some(e) ⇒ e.asInstanceOf[EffectImpl[Event, State]]
+        }
```

Comment:
On the multi-dc side we have a special type for this, `CommandHandlerImpl` but I think it is just to have a common type for both Java and Scala which isn't applicable here.